base_layer/core/src/consensus/consensus_encoding/hashing.rs

```diff
@@ -61,13 +58,12 @@ where D: Digest<OutputSize = U32>
         self.writer.0.finalize().into()
     }
 
-    pub fn update_consensus_encode<T: ConsensusEncoding + ?Sized>(&mut self, data: &T) {
-        // UNWRAP: ConsensusEncode MUST only error if the writer errors, HashWriter::write is infallible
-        data.consensus_encode(&mut self.writer)
-            .expect("Incorrect implementation of ConsensusEncoding encountered. Implementations MUST be infallible.");
+    pub fn update_consensus_encode<T: BorshSerialize>(&mut self, data: &T) {
+        BorshSerialize::serialize(data, &mut self.writer)
+            .expect("Incorrect implementation of BorshSerialize encountered. Implementations MUST be infallible.");
```

Comment:
Unfortunately, this is not guaranteed by Borsh e.g. https://github.com/near/borsh-rs/blob/master/borsh/src/ser/mod.rs#L230. I believe the only case where this can fail in practice is if you have a vec/list of length > u32::MAX but we no longer have the guarantees that ConsensusEncoding had.

A hasher that can fail is pretty inconvenient `block_header.hash().map_err(...)` but that might be the only way to do it. 

I think you can leave the expect in for now, and we'll need to make another refactor for the hash() methods